h/services/jwt.py

```diff
@@ -47,7 +57,156 @@ def decode_token(
+        method for encoding JWTs that is general enough to be applicable in
+        many contexts (whenever it would be appropriate to use a
+        symmetrically-signed but not encrypted JWT with an expiration time,
+        issuer and audience) and that uses good practices, for example:
+
+        * Requiring the JWT to have an expiration time, issuer and audience and
+          always verifying these when decoding.
+        * Signing all JWTs using a single key that is managed by JWTService, to
+          avoid a proliferation of different JWT signing keys in envvars or
+          databases. The convenience of a single key is judged worth the
+          potential loss of security.
+        * Using dataclasses to define the expected payload formats,
+          and handling exceptions when invalid payloads are encountered.
+        * Having thorough documentation, type annotations, and unittests.
+        * Providing a simple, easy to use, and easy to test interface for any
+          code that wants to encode or decode JWTs.
+
+        """
+        payload_dict = asdict(payload)
+
+        if _algorithm == "none":
+            signing_key = None
+        elif _signing_key:
+            signing_key = _signing_key
+        else:
+            signing_key = self.jwt_signing_key
+
+        # This is deliberately not mentioned in the docstring or type
+        # annotations above because it's not intended to be part of the public
+        # interface:
+        # If falsey values (e.g. None) are passed for expiration_time, issuer
+        # or audience they will be omitted. The resulting JWT will fail to
+        # decode with the decode_symmetric() method below.
+        # This is to allow unittests to generate invalid JWTs.
+        if expiration_time:
+            payload_dict["exp"] = expiration_time
+        if issuer:
+            payload_dict["iss"] = issuer
+        if audience:
+            payload_dict["aud"] = audience
+
+        return jwt.encode(
+            payload_dict,
+            key=signing_key,  # type: ignore[arg-type]
+                              # PyJWT's type annotation is wrong: it's actually
+                              # possible and documented (and sometimes even
+                              # required) to pass key=None but the type
+                              # annotation doesn't allow it.
+            algorithm=_algorithm,
+        )  # fmt: skip
+
+    # The type of the payload_class argument below is any dataclass and the
+    # return type of the decode_symmetric() function is an instance of whatever
+    # dataclass was passed as the payload_class argument.
+    # Use a TypeVar in order to correctly annotate these.
+    ANY_DATACLASS = TypeVar("ANY_DATACLASS")
+
+    def decode_symmetric(
+        self,
+        token: str,
+        *,
+        audience: str | Iterable[str],
+        payload_class: type[ANY_DATACLASS],
+    ) -> ANY_DATACLASS:
```

Comment:
Decoding does enforce that the JWT must contain an `iss` (issuer) string, and there's a unittest, but it _does not_ verify the value of the `iss` string. Hence there's no `issuer` argument here. For the reasoning see: https://github.com/hypothesis/h/pull/9738/commits/c6831148e8f4591a7a62ec8e4c8f414f782c38f7